Fait par :  
HAO Yuan 21117163 SAR  
YIFENG Yu 21113616 SAR  
XUCHEN wei 28620150 SAR  

In [2]:
import projet_optimisation

# *Préliminaire*
## *les fonctions qui peuvent etre sutilisées pour ce projet*
projet_optimisation.résultat_iteration_gradient(X,l,theta,alpha,nmax)      
--->cette fonction est pour tracer le resultat d'iteration de la methode gradient a pas fixé a partir d'un point départ choisi dans le graphe 

projet_optimisation.résultat_iteration_newton(X,l,theta,precision,nmax)    
--->cette fonction est pour tracer le resultat d'iteration de la methode newton a partir d'un point départ choisi dans le graphe 

projet_optimisation.solution_articulaire_1ere(X,l)                         
--->méthode scipy.optimize.root

projet_optimisation.solution_articulaire_2eme(X,l)                         
--->méthode scipy.optimize.minimize 

projet_optimisation.solution_articulaire_3eme(X,l,alpha,nmax)              
--->méthode grandient a pas fix 

projet_optimisation.solution_articulaire_4eme(X,l,precision,nmax)          
--->méthoode newton

projet_optimisation.solution_articulaire_analytique(X,l)                   
--->méthode analytique 

projet_optimisation.solution_et_trajectoire(X,l)                           
--->on trace le trajectoire pour les deux longueurs l1,l2 avec les esultat de premiers methode  

projet_optimisation.solution_et_trajectoire_avec_resultat_iteration_newton(X,l,precision,nmax) ---on trace le trajectoire pour les deux longueurs l1,l2 avec les resultat de quatrième methode  

## *les commentaires sur les paramètres des fonctions:*
X--->coordonnées de la pose ex:[2,3]

l--->paramètre des longeur de bras ex: l--->[l1,l2]--->[2,3]

précision--->la critière d'arret de la convergence, veuillez choisir 1e-2 pour la méthode newton pour éviter les erreurs de "nan" du python

alpha--->le pas chois pour la méhthode gradient a pas fixé, veuillez choisir 1e-2 comme la pas max, pour le cas alpha qui est plus grand que ca, on va le discuter après

nmax--->nombre max d'itération，veuilez choisir 1000, vous pouvez également choisir les autre valures, mais pas trop petite

## *Atention!* 
1.La compilation de la  methode gradient ou newton peut prendre jusqu'a une miminutes  
2.Pour exécuter le programme et afficher le résultat voulu (méthode de gradient et newton), il faut saisir les valeurs demandées  
3.Pour utiliser les fonctions, veuillez décommter les commentaires  


# *Ⅰ.Introduction:*
Dans ce TP, on a étudié le trajectoire d’un bras de robot à 2 articulations,on a appris quatre méthodes différents.Les quatre méthodes utilisent différentes méthodes d'itération. Pour chaque méthode, ses résultats itératifs sont basés sur les quelques paramètres que nous avons donnés. Afin de pouvoir trouver autant que possible des solutions possibles dans l'espace, nous avons défini de nombreux points de départ à l'intérieur de chaque méthode,pour la méthode scipy.optimize.root, on a mis 49 points départs, pour la méthode scipy.optimize.minimize, on a mis 100 points départ, pour la méthode de gradient et newton, on a  mis 400 points départ.En même temps, nous avons créé des boucles qui peuvent filtrer ces points à l'intérieur du programme, afin que notre programme puisse obtenir des résultats appropriés(solution unique, solution non unique, pas de solution) sous n'importe quelle valeur de paramètre donnée(X,l).De plus, pour les troisième et quatrième méthodes, vous pouvez choisir d'afficher ou non les résultats d'itération de quatre cents points de départ, et la situation de filtrage pour ces quatre cents points. Comme mentionné dans la préliminaire, notre programme se compose de neuf fonctions, chacune implémentant une fonction différente, et nous utilisons ces fonctions dans le contenu suivant pour illustrer les résultats que nous obtenons.



# *Ⅱ.PARTIE A*
### *1.Dans cette méthode on a tester 49 point des points départ en utilisant optimize.root, et puis on donne une méthode de fitrer ces résultats pour distinguer le cas pas de solution,solution unique et solution non-unique*
#### vou pouvez vérifier les trois cas en décommentant les fonctions suivantes

In [2]:
#vou pouvez vérifier les trois cas en décommentant les fonctions suivantes
##solution non-unique
#projet_optimisation.solution_articulaire_1ere([2,4],[2,3])
#projet_optimisation.solution_articulaire_1ere([5,9],[3,10])
#solution unique
projet_optimisation.solution_articulaire_1ere([3,4],[2,3])
projet_optimisation.solution_articulaire_1ere([5,12],[3,10])
#pas de soulution
#projet_optimisation.solution_articulaire_1ere([16,4],[2,3])
#projet_optimisation.solution_articulaire_1ere([5,13],[3,10])

la methode par optimize.root

les points départ différents pour trouver solution possible:
49 points différent choisi:
[3, -3], [3, -2], [3, -1], [3, 0], [3, 1], [3, 2], [3, 3], [2, -3], [2, -2], [2, -1], [2, 0],[2, 1], [2, 2], [2, 3], [1, -3], [1, -2], [1, -1], [1, 0], [1, 1], [1, 2], [1, 3], [0, -3], [0, -2], [0, -1], [0, 0], [0, 1], [0, 2], [0, 3], [-1, -3], [-1, -2], [-1, -1], [-1, 0], [-1, 1], [-1, 2], [-1, 3], [-2, -3], [-2, -2], [-2, -1], [-2, 0], [-2, 1], [-2, 2], [-2, 3], [-3, -3], [-3, -2], [-3, -1], [-3, 0], [-3, 1], [-3, 2], [-3, 3]

les résultats résolu par root en 49 points départ:
theta1----->
 [0.927295, 0.927295, 0.927295, 0.927306, 0.927294, 0.927295, 0.927295, 0.927296, 0.927295, 0.927295, 0.927288, 0.927295, 0.927295, 0.927294, 0.927297, 0.927295, 0.927295, 0.927295, 0.927295, 0.927295, 0.927295, 0.927296, 0.927296, 0.927295, 0.927295, 0.927295, 0.927295, 0.927295, 0.927295, 0.927295, 0.927296, 0.928081, 0.927295, 0.927295, 0.927295, 0.927295, 0.927295, 0.927295, 0.

(array([0.        , 0.0240001 , 0.0480002 , 0.07200031, 0.09600041,
        0.12000051, 0.14400061, 0.16800071, 0.19200082, 0.21600092,
        0.24000102, 0.26400112, 0.28800122, 0.31200133, 0.33600143,
        0.36000153, 0.38400163, 0.40800173, 0.43200184, 0.45600194,
        0.48000204, 0.50400214, 0.52800224, 0.55200235, 0.57600245,
        0.60000255, 0.62400265, 0.64800276, 0.67200286, 0.69600296,
        0.72000306, 0.74400316, 0.76800327, 0.79200337, 0.81600347,
        0.84000357, 0.86400367, 0.88800378, 0.91200388, 0.93600398,
        0.96000408, 0.98400418, 1.00800429, 1.03200439, 1.05600449,
        1.08000459, 1.10400469, 1.1280048 , 1.1520049 , 1.176005  ]),
 array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),
 [],
 [],
 True,
 False)

### *2.Dans cette partie, on a fait le programme de tracer le trajectoire en utilisant le résutat du méthode de optimize.root et la méthode newton*
Pour le trajectoire en utilisant le résutat du méthode de optimize.root, l'idée est qu'on donne un coordonnée de la pose et la paramètre de la longueur des bras, et d'après le méthode root, on a un résulatat pour theta1 et theta2 et on défini un temps passé (50), on divise la valeur de theta par 50 en utilisant linspace, et on utilise le list de ces deux linspace pour tracer le tracectoire,en générale, il y en a trois cas: pas de solution, solution unique et solution non-unique,le cas pas de solution et solution unique, on va tracer une tracjectoire et pour le cas solution non-unique, on va tracer deux trajectoire en un graphe  
Et pour le trajectoire en utilisant le résutat du méthode newton, tout d'abord, on va compiler solution_articulaire_4eme pour trouver les points départ les plus approprié,trois solutions possibles, soit une point départ qui est correspondant au cas pas de solution ou solution unique, soit deux points départs qui est correspondant au cas solution non_unique.ensuite, on va utiliser ces points départ dans la fonction Newton(theta,precision,nmax) pour faire les itération, En itérant, nous obtiendrons deux listes en même temps, correspondant à theta1 et theta2,et puis on trace le trajectoire du robot a l'instant différent d'apres ces deux tableau. 

#### vou pouvez vérifier les trois cas en décommentant les fonctions suivantes

In [ ]:
#la méthode optimize.root
##solution non-unique
###projet_optimisation.solution_et_trajectoire([2,4],[2,3])
projet_optimisation.solution_et_trajectoire([5,9],[3,10])
##solution unique
###projet_optimisation.solution_et_trajectoire([3,4],[2,3])
###projet_optimisation.solution_et_trajectoire([5,12],[3,10])
##pas de solution
###projet_optimisation.solution_et_trajectoire([16,4],[2,3])
###projet_optimisation.solution_et_trajectoire([5,13],[3,10])

#la méthode newton
##solution non-unique
###projet_optimisation.solution_et_trajectoire_avec_resultat_iteration_newton([1,4],[2,3],1e-2,1000)
projet_optimisation.solution_et_trajectoire_avec_resultat_iteration_newton([3,9],[3,10],1e-2,1000)
##solution unique
###projet_optimisation.solution_et_trajectoire_avec_resultat_iteration_newton([3,4],[2,3],1e-2,1000)
###projet_optimisation.solution_et_trajectoire_avec_resultat_iteration_newton([5,12],[3,10],1e-2,1000)
##pas de solution
###projet_optimisation.solution_et_trajectoire_avec_resultat_iteration_newton([3,5],[2,3],1e-2,1000)
###projet_optimisation.solution_et_trajectoire_avec_resultat_iteration_newton([5,13],[3,10],1e-2,1000)

la methode par optimize.root

les points départ différents pour trouver solution possible:
49 points différent choisi:
[3, -3], [3, -2], [3, -1], [3, 0], [3, 1], [3, 2], [3, 3], [2, -3], [2, -2], [2, -1], [2, 0],[2, 1], [2, 2], [2, 3], [1, -3], [1, -2], [1, -1], [1, 0], [1, 1], [1, 2], [1, 3], [0, -3], [0, -2], [0, -1], [0, 0], [0, 1], [0, 2], [0, 3], [-1, -3], [-1, -2], [-1, -1], [-1, 0], [-1, 1], [-1, 2], [-1, 3], [-2, -3], [-2, -2], [-2, -1], [-2, 0], [-2, 1], [-2, 2], [-2, 3], [-3, -3], [-3, -2], [-3, -1], [-3, 0], [-3, 1], [-3, 2], [-3, 3]

les résultats résolu par root en 49 points départ:
theta1----->
 [2.389221, 2.389221, 2.389221, 2.389218, 6.02136, 6.02136, 2.389221, 6.021361, 2.389221, 2.389221, 2.389221, 2.389221, 2.38922, 6.021358, 6.02136, 2.389221, 6.02136, 6.02136, 6.02136, 2.38922, 6.02136, 6.021361, 6.02136, 6.02136, 6.02136, 6.02136, 6.02136, 2.38922, 2.389221, 6.02136, 2.389221, 2.389223, 6.02136, 6.02136, 2.389221, 6.02136, 6.02136, 6.02136, 2.38921, 6.02136, 2.389

### *3.Deuxième méthode : utiliser la fonction « minimize » de la bibliothèque scypy.optimize appliquées au carré de la norme du résidu.*
#### Dans cette méthode on a tester 100 points des point départ en utilisant optimize.minimize, et puis on donne un méthode de fitrer ces résultats pour distinguer le cas pas de solution,solution unique et solution non-unique
#### vou pouvez vérifier les trois cas en décommentant les fonctions suivantes

In [3]:
##solution non-unique
###projet_optimisation.solution_articulaire_2eme([5,8],[3,10])
projet_optimisation.solution_articulaire_2eme([2,4],[2,3])
##solution unique
###projet_optimisation.solution_articulaire_2eme([5,12],[3,10])
projet_optimisation.solution_articulaire_2eme([3,4],[2,3])
##pas de solution
###projet_optimisation.solution_articulaire_2eme([5,20],[3,10])
###projet_optimisation.solution_articulaire_2eme([3,5],[2,3])

la methode par optimize.minimize


les résultats résolu par minimize en 100 points départ:
theta1----->
 [0.530924, 0.530924, 0.530924, 0.530924, 0.530924, 0.530924, 1.683374, 1.683375, 0.530923, 0.530924, 0.530924, 0.530924, 0.530924, 0.530924, 0.530924, 0.530923, 1.683373, 1.683374, 1.683373, 0.530924, 1.683374, 0.530924, 0.530924, 0.530924, 0.530923, 0.530924, 1.683374, 1.683374, 1.683374, 1.683374, 1.683373, 1.683374, 0.530924, 0.530924, 1.683374, 1.683374, 1.683374, 1.683374, 1.683374, 1.683373, 1.683374, 1.683372, 1.683374, 1.683372, 1.683374, 1.683374, 1.683374, 1.683375, 1.683374, 1.683374, 0.530924, 1.683373, 1.683374, 1.683374, 1.683374, 1.683374, 1.683374, 1.683374, 1.683374, 0.530924, 1.683373, 1.683374, 0.530924, 0.530923, 0.530924, 1.683376, 1.683374, 1.683372, 0.530924, 1.683373, 1.683374, 0.530924, 0.530924, 0.530924, 0.530924, 0.530924, 1.683374, 1.683373, 0.530926, 1.683374, 0.530924, 0.530924, 0.530923, 0.530924, 0.530924, 0.530925, 0.530924, 1.683375, 0.530924, 0.53

### *4.Troisième méthode : écrire le programme pour minimiser le carré de la norme du résidu en appliquant la méthode du gradient à pas fixe, avec un mécanisme qui garantit que la norme du résidu diminue à chaque itération*
Dans cette méthode on a tester 400 points des point départ en utilisant gradient a pas fixé, et puis on donne un méthode de fitrer ces résultats pour distinguer le cas pas de solution,solution unique et solution non-unique,Le programme trouvera la solution correspondante parmi ces quatre cents points, s'il y a une solution, il trouvera une solution optimale et son point de départ correspondant(Nous avons écrit une méthode précise dans le programme pour suivre le point de départ, c'est-à-dire que nous pouvons clairement savoir quel est le point de départ correspondant à la solution optimale(theta1,theta2)obtenue par le programme), s'il y a deux solution, il trouvera deux solution optimale et son point de départ correspondants, s'il y a pas de solution, il trouvera une  solution qui est le plus proche a la pose donnée.
#### vou pouvez vérifier les trois cas en décommentant les fonctions suivantes

In [4]:
##solution non-unique
###projet_optimisation.solution_articulaire_3eme([1,4],[2,3],0.01,1000)
projet_optimisation.solution_articulaire_3eme([3,9],[3,10],0.001,1000)
##solution unique
###projet_optimisation.solution_articulaire_3eme([3,4],[2,3],0.01,1000)
###projet_optimisation.solution_articulaire_3eme([5,12],[3,10],0.001,1000)
##pas de solution
###projet_optimisation.solution_articulaire_3eme([3,5],[2,3],0.01,1000)
###projet_optimisation.solution_articulaire_3eme([5,13],[3,10],0.001,1000)

la méthode de gradient a pas fixé avec alpha choisi---> 0.001  et le nombre d'itération max---> 1000

veuillez patienter, les donneés en train de chargement, la compilation peut prendre jusqu'a trois minute


veuillez taper 'yes' si vous voulez afficher les resultats de methode gradient a pas fix sinon taper'no'---> yes



les résultats résolu par methode gradient a pas fixe en 400 points départ:

theta----->
 [[2.837411, 4.390176], [2.837412, 4.390175], [5.943867, 1.893008], [2.837408, 4.390179], [5.943863, 1.893011], [5.943861, 1.893014], [5.943861, 1.893014], [2.837406, 4.39018], [2.837405, 4.390182], [5.943859, 1.893016], [2.837404, 4.390183], [2.837403, 4.390184], [5.943857, 1.893017], [2.83742, 4.390167], [2.837402, 4.390185], [5.943855, 1.893019], [2.8374, 4.390186], [5.943877, 1.892999], [5.943853, 1.893022], [2.837425, 4.390163], [2.837398, 4.390189], [5.943851, 1.893023], [5.943882, 1.892994], [2.837428, 4.39016], [5.943883, 1.892993], [5.943848, 1.893026], [5.943884, 1.892992], [5.943884, 1.892992], [5.943847, 1.893027], [2.83743, 4.390158], [5.943885, 1.892991], [5.943885, 1.892991], [5.943885, 1.892991], [5.943885, 1.892991], [5.943886, 1.89299], [2.837433, 4.390155], [5.943888, 1.892988], [5.943842, 1.893032], [2.837436, 4.390152], [2.837438, 4.39015], [5.943837, 1.893037], [2.83744, 4.390

### *5.Quatrième méthode : écrire le programme pour minimiser le carré de la norme du résidu en appliquant la méthode de Newton.*
Dans cette méthode on a tester 400 points des point départ en utilisant méthode, et puis on donne un méthode de fitrer ces résultats pour distinguer le cas pas de solution,solution unique et solution non-unique, pareil comme la méthode gradient,s'il y a une solution, il trouvera une solution optimale et son point de départ correspondant(Nous avons écrit une méthode précise dans le programme pour suivre le point de départ, c'est-à-dire que nous pouvons clairement savoir quel est le point de départ correspondant à la solution optimale(theta1,theta2)obtenue par le programme), s'il y a deux solution, il trouvera deux solution optimale et son point de départ correspondants, s'il y a pas de solution, il trouvera une  solution qui est le plus proche a la pose donnée.
#### vou pouvez vérifier les trois cas en décommentant les fonctions suivantes

In [6]:
##solution non-unique
projet_optimisation.solution_articulaire_4eme([1,4],[2,3],1e-2,1000)
###projet_optimisation.solution_articulaire_4eme([3,9],[3,10],1e-2,1000)
##solution unique
###projet_optimisation.solution_articulaire_4eme([3,4],[2,3],1e-2,1000)
###projet_optimisation.solution_articulaire_4eme([5,12],[3,10],1e-2,1000)
##pas de solution
###projet_optimisation.solution_articulaire_4eme([3,5],[2,3],1e-2,1000)
###projet_optimisation.solution_articulaire_4eme([5,13],[3,10],1e-2,1000)

la methode de newton avec la précision choisi---> 0.01 et le nombre d'itération max---> 1000

veuillez patienter, les donneés en train de chargement, la compilation peut prendre jusqu'a trois minute


veuillez taper 'yes' si vous voulez afficher les resultats de methode newton sinon taper 'no'---> yes



les résultats résolu par methode newton en 400 points départ:

theta----->
 [[2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [0.569848, 1.23096], [2.081788, 5.052225], [2.081788, 5.052225], [0.569849, 1.230959], [0.569849, 1.230959], [0.569849, 1.230959], [0.569849, 1.230959], [0.569848, 1.230959], [0.569848, 1.230959], [0.569848, 1.230959], [0.569848, 1.230959], [0.569848, 1.230959], [0.569848, 1

[[0.6613879270715354, 2.6455517082861415], [0.0, 0.3306939635357677]]

### *6.Le méthode analytique* 
Cette méthode est realisé par le méthode de root et les formules calculé manuellement pour le cas solution non-unique, on utilise la méthode root pour obtenir une solution, et puis on calculer une autre solution directement par le formule calculé,ce méthode est le plus précis, il y a pas de posiibilité d'ommetre les solution, si la pose donnée est bien défini dans l'espace atteignable non-unique, on aura certainement deux solutions,cependant on peux constater quelque  inconvénient de cette méthode, si le systète robotique a résoudre est assez complex, on ne peut pas trouver les solution par la méthode mathematique,parce que les formule sera tres complex ou impossible de les trouver!
#### vou pouvez vérifier les trois cas en décommentant les fonctions suivantes

In [7]:
##solution non-unique
projet_optimisation.solution_articulaire_analytique([1,4],[2,3])
##solution unique
###projet_optimisation.solution_articulaire_analytique([3,4],[2,3])
##pas de solution
###projet_optimisation.solution_articulaire_analytique([1,15],[2,3])

la methode anamytique

 solution non-uniques:
 la première solution: [0.5698482532666862, 1.2309594173060567] 
 l'autre solution : [2.0818794729463552, 5.052174549854371]

 vérification de la valeur:
 pour theta,modèle directe nous donne: [1.0000000000111653, 4.000000000046308] 
 pour theta_prime,modèle directe nous donne: [0.9997462033555296, 3.9999908347378197]
en génerale, on utilise le symétrie pour trouver une autre solution,les deux solition sont symétrique par rappor a la ligne OP(O est le point origine et P est le point de la pose)


([0.5698482532666862, 1.2309594173060567],
 [2.0818794729463552, 5.052174549854371])

# *Ⅲ.PARTIE B*
## *Importance du pas de la méthode du gradient*
Apres d'avoir compilé les deux fonction suivante avec deux alpha différent(0.01 ou 0.05), on peut constater, le résumtat trouvé par la méthode gradient n'est pas paraeil
voici sont les résultat:
alpha--->0.05:  
le cas pas de solution,le point le plus proche: point depart( [5.291103416572283, 5.9524913436438185] )   --->theta: [0.010059, 1.236812] et la distance le plus court: 2.2608399890116297  
alpha--->0.01  
les solutions posisbles trouvées---> [[2.081787, 5.052226], [0.569848, 1.23096]]   
solution non_uniques!  
premiere solution nous donne---> point depart( [0.0, 4.6297154895007475] )---> [2.081787, 5.052226]   vérification par modèle geometrique---> [1.0000000479014992, 4.000000143816065]  
deuxieme solution nous donne---> point depart( [0.0, 5.9524913436438185] )---> [0.569848, 1.23096]   vérification par modèle geometrique---> [0.9999993110338806, 3.9999993482367535]    
Particulierement, on peut vérifier l'importance du choix de bon alpha par la fonction "résultat_iteration_gradient(X,l,theta,precision,nmax)"
en compilant les deux exemple suivantes, on peut constater quand on choisi 0.05 comme alpha, il ne peut pas coverger correctement  
### *Petite conclusion:*     
choisir une bonne valeur pour le pas de la méthode gradient est tres important, si le alpha est trop grand, la methode ne peut pas converger correctement, mais si le alpha est trop petit, la rapidité de la compilation est top bas, 
donc pour avir un résultat une bonne rapidité, on doit choisir une pertinent alpha.
#### vou pouvez les vérifier en décommentant les fonctions suivantes

In [8]:
projet_optimisation.solution_articulaire_3eme([1,4],[2,3],0.01,1000)
#projet_optimisation.solution_articulaire_3eme([1,4],[2,3],0.05,1000)
projet_optimisation.résultat_iteration_gradient([1,4],[2,3],[0,0],0.01,1000) 
#projet_optimisation.résultat_iteration_gradient([1,4],[2,3],[0,0],0.05,1000) 

la méthode de gradient a pas fixé avec alpha choisi---> 0.01  et le nombre d'itération max---> 1000

veuillez patienter, les donneés en train de chargement, la compilation peut prendre jusqu'a trois minute


veuillez taper 'yes' si vous voulez afficher les resultats de methode gradient a pas fix sinon taper'no'---> yes



les résultats résolu par methode gradient a pas fixe en 400 points départ:

theta----->
 [[2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.081787, 5.052226], [2.0817

# *Ⅳ.PARTIE C*
## *Importance du point de départ de la méthode de Newton*
On peut utiliser la fontion "projet_optimisation.résultat_iteration_newton(X,l,theta,precision,nmax)"pour vérifier l'importance du point de départ de la méthode newton  
quand on choisi [5,1.25] comme le point départ, on peut constater que la fonction est convergé en un point ni point max ni point max(cette fonction converge vers l'extremum local)  
quand on choisi [5,0] comme le point départ, on peut constater que la fonction est convergé en max   
quand on choisi [1.25,2.5] comme le point départ, on peut constater que la fonction est convergé en min  
### *Petite conclusiton:* 
le choix différent du point départ de la méthode newton influence beaucoup le résultat de la méthode newton, pour avoir un résultat correct, on doit tester plein de points départs, et puis on doit filtrer ces résultat si bine que nous puissons avoir un résultat correct
#### vou pouvez les vérifier en décommentant les fonctions suivantes

In [9]:
projet_optimisation.résultat_iteration_newton([1,4],[2,3],[5,1.25],0.01,1000) 
#projet_optimisation.résultat_iteration_newton([1,4],[2,3],[5,0],0.01,1000) 
#projet_optimisation.résultat_iteration_newton([1,4],[2,3],[1.25,2.5],0.01,1000) 

# *Ⅴ.Conclusion finale:*
Dans ce projet, on a programmé quatre méthodes pour trouve la solution de la modèle inverse d’un bras de robot à 2 articulations, methode optimize.root, méthode.optimize, methode gradient a pas fixé, méthode newton,en fait ,la pricipe de ces methodes différentes sont meme, on donne un point départ, et puis les méthode va trouve une valeur minimun local apres la convergence, mais Ces méthodes ne peuvent produire qu'un seul résultat à la fois，Le résultat obtenu à chaque fois n'est pas forcément la réponse que l'on souhaite, Afin de considérer toutes les situations(solution unique, non-unique,pas de solution) et de trouver la plus petite valeur, nous devons tester différents points de départ. Nous avons ajouté une liste de points de départ pour chaque méthode et écrit une boucle pour tester chaque point de départ.Nous avons agrégé tous les résultats obtenus, et fait un tri, ajouté quelques restrictions(par exemple:minimisation de la valeur articulaire, enlever les valeurs équivalents), et enfin notre programme peut distinguer quelle situation est à partir des résultats obtenus(solution unique, non-unique,pas de solution).  
Pour ces quatre méthodes,on constater que nous avons constaté que les première et deuxième méthodes obtiennent les résultats souhaités sans avoir à tester de nombreux points de départ(root,minimize), Les troisième et quatrième méthodes nécessitent davantage de données de départ pour le dépistage(gradient a pas fixé,newton),pour la methode gradient a pas fixé, choix du bonne valeur du pas est tres important, si alpha est tres grand, la convergence de cette méthode n'est pas tres correcte, on aura pas un correct résultat, et pour la méthode newton, le choix du point départ est tres important, Si le point de départ que nous choisissons ne convient pas, alors cette fonction peut converger vers l'extremum local.Et on constate que la méthode newton converge plus vite que la méthode de gradient a pas fixé.     
Nous avons beaucoup appris grâce à ce projet. Non seulement dans le domaine de la robotique, mais aussi dans un large éventail de domaines d'ingénierie, l'optimisation est un outil nécessaire. Souvent, nous devons trouver la solution optimale, mais pour certaines situations très compliquées, trouver la solution optimale pas si facile  
En fait, il existe de nombreux algorithmes d'optimisation, nous devons maîtriser davantage de méthodes d'optimisation et être en mesure de les appliquer dans des situations appropriées，par ailleur, nous avons aussi essyé utilisé la méthode relaxaion pour résoudre la problème demandé, 
mais malheureusement, nous n'avons pas de le résussir.
